channel/channel.ml

```diff
@@ -50,66 +52,82 @@ module Make(Flow:V1_LWT.FLOW) = struct
   let ibuf_refill t =
     Flow.read t.flow >>= function
     | `Ok buf ->
+      (* users of get_ibuf (and therefore ibuf_refill) expect the buffer
+         returned here to have length >0; if Flow.read ever gives us empty
+         buffers, this will be violated causing Channel users to see Cstruct
+         exceptions *)
       t.ibuf <- Some buf;
       return_unit
-    | `Error _ | `Eof ->
-      fail Closed
+    | `Error e ->
+      fail (Read_error (Flow.error_message e))
+    | `Eof ->
+      (* close the flow before throwing exception; otherwise it will never be
+         GC'd *)
```

Comment:
The flow provided by `tcp/flow.ml` (and `tcp/pcb.ml`) doesn't do anything to clean itself up on EOF; rather it relies on the state machine logic (`tcp/state.ml`) to call `close` when logic dictates that the communications are over.  I suspect this particular case is related to https://github.com/mirage/mirage-tcpip/issues/107 (we don't close Established connections when we receive an RST).